### ISSUE TITLE
Seeing weird occasional response times

### ISSUE THREAD

**Issue by OskarKjellin** [ReactiveX/RxNetty#382]
Hi,

I'm seeing some response times that cannot be expected (around 1 second occasionally). Started seeing these on our regular tests with our services even with really low concurrent requests. Bumping up the number of concurrent requests even above 10 starts to give a lot of timeouts. 

I managed to make a small example to reproduce it and I see the high response times even with only 1 thread sending requests. Am I using it completely wrong or is this expected? 

Small test program:

```
  public static void main(String[] args) throws Exception
    {
        HttpClient<ByteBuf, ByteBuf> rxNetty = RxNetty.createHttpClient("127.0.0.1", 1337);

        HttpServer<ByteBuf, ByteBuf> server = RxNetty.createHttpServer(1337, (HttpServerRequest<ByteBuf> request, HttpServerResponse<ByteBuf> response) -> {
            response.setStatus(HttpResponseStatus.OK);
            response.writeString("OK");
            return response.close();
        });
        server.start();
        //Just set the logging to INFO to not log too much
        LoggerContext ctx = (LoggerContext) LogManager.getContext(false);
        Configuration config = ctx.getConfiguration();
        LoggerConfig loggerConfig = config.getLoggerConfig(LogManager.ROOT_LOGGER_NAME);
        loggerConfig.setLevel(Level.INFO);
        ctx.updateLoggers();

        List<Thread> threads = new ArrayList<>();
        int threadCount = 1;
        for (int i = 0; i < threadCount; i++)
        {
            Runnable runnable = new Runnable()
            {
                @Override
                public void run()
                {
                    for (int x = 0; x < 10000; x++)
                    {
                        try
                        {
                            long start = System.currentTimeMillis();
                            HttpClientRequest<ByteBuf> httpRequest = HttpClientRequest.createGet("/test");

                            Observable<String> observable = rxNetty.submit(httpRequest)
                                    .flatMap((HttpClientResponse<ByteBuf> r) ->
                                            r.getContent().map(b -> b.toString(StandardCharsets.UTF_8))
                                    );
                            String response = observable.toBlocking().last();
                            long dur = System.currentTimeMillis() - start;
                            if (dur > 100)
                            {
                                System.out.println(x + ":" + dur);
                            }
                        }
                        catch (Exception ex)
                        {
                            ex.printStackTrace();
                        }
                    }
                }
            };
            Thread t = new Thread(runnable);
            t.start();
            threads.add(t);
        }
        for (Thread t : threads)
        {
            t.join();
        }
        server.shutdown();
    }
```

**Comment by NiteshKant** [comment]
@OskarKjellin If you change your code to not use `.toBlocking()` you will see that the response times are much lower. 

``` java

HttpClient<ByteBuf, ByteBuf> rxNetty = RxNetty.createHttpClient("127.0.0.1", 1337);

HttpServer<ByteBuf, ByteBuf> server = RxNetty.createHttpServer(1337,
                                                               (HttpServerRequest<ByteBuf> request, HttpServerResponse<ByteBuf> response) -> {
                                                                   response.setStatus(HttpResponseStatus.OK);
                                                                   response.writeString("OK");
                                                                   return response.close();
                                                               });
server.start();

List<Thread> threads = new ArrayList<>();
int threadCount = 1;
for (int i = 0; i < threadCount; i++)
{
    Runnable runnable = () -> {
        for (int x = 0; x < 10000; x++)
        {
            try
            {
                long start = System.currentTimeMillis();
                HttpClientRequest<ByteBuf> httpRequest = HttpClientRequest.createGet("/test");

                final int finalX = x;
                rxNetty.submit(httpRequest)
                       .flatMap((HttpClientResponse<ByteBuf> r) ->
                                        r.getContent().map(b -> b.toString(
                                                StandardCharsets.UTF_8))
                       )
                       .doOnCompleted(() -> {
                           long dur = System.currentTimeMillis() - start;
                           if (dur > 400) {
                               System.out.println(finalX + ":" + dur);
                           }
                       })
                       .subscribe();
            }
            catch (Exception ex)
            {
                ex.printStackTrace();
            }
        }
    };
    Thread t = new Thread(runnable);
    t.start();
    threads.add(t);
}
for (Thread t : threads)
{
    t.join();
}
server.shutdown();
```

By using `.toBlocking()` you are only sending one request at a time and hence not utilizing the parallelism in the system ( > 1 eventloops). 

I am closing this issue, if you think it is still an issue, feel free to reopen.

**Comment by OskarKjellin** [comment]
@NiteshKant I will check more in detail tomorrow. I am not using `toBlocking()` in the original place where I found this so I will need to find out another example not using  `toBlocking()` that can reproduce it then.

**Comment by OskarKjellin** [comment]
Even if I am using toBlocking and sending 1 request at a time, should I still be getting that high response times? I mean, 1 concurrent request should never lead to response times as what I am seeing (around 1 second)

**Comment by OskarKjellin** [comment]
@NiteshKant I think your test is quite broken unfortunately. It does not wait for all the requests to arrive before it quits and it still shows very high response times (I think over 1 second for local requests is very high?). It's funny though. Adding more threads sending requests seem to provide better response times. But I am still wondering why my single thread tests are performing so badly. 

By the way please reopen the ticket as I don't seem to have the rights to do so.

``` java
   HttpClient<ByteBuf, ByteBuf> rxNetty = RxNetty.createHttpClient("127.0.0.1", 1337);

        HttpServer<ByteBuf, ByteBuf> server = RxNetty.createHttpServer(1337, (HttpServerRequest<ByteBuf> request, HttpServerResponse<ByteBuf> response) -> {
            response.setStatus(HttpResponseStatus.OK);
            try
            {
                //Simulate network
                Thread.sleep(10);
            }
            catch (InterruptedException e)
            {
                e.printStackTrace();
            }
            response.writeString("OK");
            return response.close();
        });
        server.start();
        //Just set the logging to INFO to not log too much
        LoggerContext ctx = (LoggerContext) LogManager.getContext(false);
        Configuration config = ctx.getConfiguration();
        LoggerConfig loggerConfig = config.getLoggerConfig(LogManager.ROOT_LOGGER_NAME);
        loggerConfig.setLevel(Level.INFO);
        ctx.updateLoggers();

        List<Thread> threads = new ArrayList<>();
        int threadCount = 1;
        for (int i = 0; i < threadCount; i++)
        {
            Runnable runnable = new Runnable()
            {
                private final AtomicInteger requests = new AtomicInteger();

                @Override
                public void run()
                {
                    sendRequest();
                    while (requests.get() < 10000)
                    {
                        try
                        {
                            Thread.sleep(1000);
                        }
                        catch (InterruptedException e)
                        {
                            e.printStackTrace();
                        }
                    }
                }

                public void sendRequest()
                {
                    int req = requests.incrementAndGet();
                    if (req > 10000)
                    {
                        return;
                    }
                    long start = System.currentTimeMillis();
                    HttpClientRequest<ByteBuf> httpRequest = HttpClientRequest.createGet("/test");

                    rxNetty.submit(httpRequest)
                            .flatMap((HttpClientResponse<ByteBuf> r) ->
                                    r.getContent().map(b -> b.toString(StandardCharsets.UTF_8))
                            ).subscribe(c -> {
                        long dur = System.currentTimeMillis() - start;
                        if (dur > 100)
                        {
                            System.out.println(req + ":" + dur);
                        }
                        sendRequest();
                    }, ex -> ex.printStackTrace());


                }
            };
            Thread t = new Thread(runnable);
            t.start();
            threads.add(t);
        }
        for (Thread t : threads)
        {
            t.join();
        }
        server.shutdown();
```

**Comment by OskarKjellin** [comment]
@NiteshKant Whenever I see high response times in our real environment there's usually a huge gap from the time we kick off the execution of the observable until netty has logged the WRITE request. Does that sound right? Only sending three concurrent requests we get 800 MS from when we observe the event until it has been written.

When logging out all responsetimes of the sample code above I get really high deviations:

```
1023:12
1024:13
1025:13
1026:754
1027:11
1028:12
1029:11
...
5129:11
5130:11
5131:12
5132:11
5133:11
5134:1006
5135:804
5136:11
5137:11
5138:11
5139:11
5140:11
..
9726:11
9727:12
9728:1000
9729:803
9730:12
9731:10
9732:11

```

**Comment by OskarKjellin** [comment]
@NiteshKant The plot thickens. It seems like they all happen within a small range of milliseconds of seconds (though an arbitrary number of seconds in between). Could it be some kind of connection that is slow and it occurs when the connection is released?

```
2015-07-01T13:25:13.415+02:00 - 92:845
2015-07-01T13:25:14.416+02:00 - 128:557
2015-07-01T13:25:16.417+02:00 - 244:602
2015-07-01T13:25:17.249+02:00 - 251:758
2015-07-01T13:25:21.250+02:00 - 511:830
2015-07-01T13:25:22.253+02:00 - 513:989
2015-07-01T13:25:24.254+02:00 - 621:743
2015-07-01T13:25:25.255+02:00 - 673:402
2015-07-01T13:25:26.423+02:00 - 702:837
2015-07-01T13:25:27.256+02:00 - 703:833
2015-07-01T13:25:29.258+02:00 - 855:199
2015-07-01T13:25:30.258+02:00 - 905:433
2015-07-01T13:25:30.423+02:00 - 906:165
2015-07-01T13:25:31.259+02:00 - 919:695
2015-07-01T13:25:33.261+02:00 - 1025:787
2015-07-01T13:25:41.431+02:00 - 1654:859
2015-07-01T13:25:44.266+02:00 - 1865:441
2015-07-01T13:26:09.445+02:00 - 4068:356
2015-07-01T13:26:18.281+02:00 - 4773:893
2015-07-01T13:26:18.448+02:00 - 4774:167
2015-07-01T13:26:19.282+02:00 - 4817:355
2015-07-01T13:26:29.454+02:00 - 5630:983
2015-07-01T13:26:38.290+02:00 - 6377:411
2015-07-01T13:26:42.292+02:00 - 6655:832
2015-07-01T13:26:42.462+02:00 - 6656:170
2015-07-01T13:26:43.462+02:00 - 6658:986
2015-07-01T13:26:53.468+02:00 - 7528:255
```

Sorry for spamming. Been at this for a couple of days and tried almost every way I find but I cannot seem to get it to perform.

**Comment by OskarKjellin** [comment]
Okay, sorry for spamming like this but I found something even more peculiar. If I change to your style of doing `doOnCompteted()`it performs as expected. Changing that to `doOnNext()` I see the same occasional high response times. 
Example code below:

This shows some very slow response times:

``` java
                    RxNetty.createHttpGet("http://localhost:1337/test?id=" + req)
                            .flatMap((HttpClientResponse<ByteBuf> r) ->
                                    r.getContent().map(b -> b.toString(StandardCharsets.UTF_8))
                            ).doOnNext(c -> {
                        long dur = System.currentTimeMillis() - start;
                        if (dur > 100)
                        {
                            logger.warn(req + ":" + dur);
                        }
                        sum += dur;
                        sendRequest();
                    }).subscribe();
```

This runs as fast as expected

``` java
                    RxNetty.createHttpGet("http://localhost:1337/test?id=" + req)
                            .flatMap((HttpClientResponse<ByteBuf> r) ->
                                    r.getContent().map(b -> b.toString(StandardCharsets.UTF_8))
                            ).doOnCompleted(() -> {
                        long dur = System.currentTimeMillis() - start;
                        if (dur > 100)
                        {
                            logger.warn(req + ":" + dur);
                        }
                        sum += dur;
                        sendRequest();
                    }).subscribe();

```

**Comment by luckyswede** [comment]
I'm having the same issue, occasional long response times. 
Using latest code.
I'm not using toBlocking

**Comment by OskarKjellin** [comment]
@NiteshKant added my own metricsEventListener to see if I can find any events when this happens. Every time i have a high time for the requests I see a big gap between two events right next to eachother. FlushStart and BytesRead:

```
2015-07-02 11:20:56,641 INFO  [rxnetty-nio-eventloop-3-4] client.LoadBalancingClientTests - FlushStart
2015-07-02 11:20:57,440 INFO  [rxnetty-nio-eventloop-3-2] client.LoadBalancingClientTests - BytesRead
```

**Comment by billyyarosh** [comment]
Can some one test that this resolved their slow response times?

Groovy code we used which resolved the occasional slow requests.

``` groovy
import io.netty.util.concurrent.SingleThreadEventExecutor

Field field = SingleThreadEventExecutor.class.getDeclaredField("SCHEDULE_PURGE_INTERVAL")
field.accessible = true
field.modifiers = Modifier.PUBLIC | Modifier.STATIC
field.set(null, TimeUnit.MILLISECONDS.toNanos(100))
```

**Comment by billyyarosh** [comment]
Feel this issue needs reopened as there is a real problem during the request phase of an httpclient.

We have experienced the same issue when using Ribbon. We traced it down and If you do a `toBlocking()` it happens more often, but we think it still happens in a pure async, but less often. That is harder to confirm since we had to add pauses between each request in the main thread.

**NioEventLoop selector.select does not get interrupted** if the write request task happens at the same.

Groovy code we used which resolved the occasional slow requests.

``` groovy
import io.netty.util.concurrent.SingleThreadEventExecutor
Field field = SingleThreadEventExecutor.class.getDeclaredField("SCHEDULE_PURGE_INTERVAL")
field.accessible = true
field.modifiers = Modifier.PUBLIC | Modifier.STATIC
field.set(null, TimeUnit.MILLISECONDS.toNanos(100))
```

I don't know enough about the Netty or the RXNetty to know the resolution, just that there seems to be some race condition occurring.

I'd like for someone in the #382 thread to confirm the above yields consistent load times for them.
